loadbalancer: Implement UDP socket termination

joamaki opened the pull request:
This ports the `pkg/service/connections.go` to the new control-plane.

joamaki committed:
loadbalancer/reconciler: Implement socket termination

This ports pkg/service/connections.go to the new control-plane to terminate
connected UDP sockets when the backend is deleted or marked unhealthy.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
netns: Add missing methods to netns_other.go

The Current() and All() were missing which makes the linter unhappy.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

tommyp1ckles commented on pkg/loadbalancer/config.go:
```diff
@@ -102,10 +108,15 @@ type LBMapsConfig struct {
 	MaglevMapMaxEntries                                             int
 }
 
+const defaultMaxSockRevNatEntries = 256 * 1024
```
Why this number in particular?

joamaki replied:
Unclear to me why this was chosen originally (https://github.com/cilium/cilium/blob/main/pkg/maps/lbmap/ipv4.go#L29)

tommyp1ckles commented on pkg/loadbalancer/reconciler/termination.go:
```diff
@@ -0,0 +1,252 @@
+func socketTerminationLoop(p socketTerminationParams, ctx context.Context, health cell.Health) error {
+	wtxn := p.DB.WriteTxn(p.Backends)
+	changeIter, err := p.Backends.Changes(wtxn)
+	wtxn.Commit()
```
why do we commit if it fails?

joamaki replied:
simpler code, nothing inserted if it fails, shouldn't really never fail.

tommyp1ckles commented on pkg/loadbalancer/reconciler/termination.go:
```diff
@@ -0,0 +1,252 @@
+			// Terminate the sockets connected to backends that have been either
+			// deleted or which are no longer considered viable.
+			if change.Deleted || !backend.IsAlive() {
+				opSupported := terminateUDPConnectionsToBackend(p, backend.Address)
```
Would it be better to just probe for this kernel functionality once and then avoid retrying this and then avoid having these error logs periodically appear?

joamaki replied:
This was how the original implemented it and didn't feel like diving into how to do this differently when porting this. This is not retrying as it's exiting the one-off job here.

jrife replied:
With `BPFSocketDestroyer`, I'm probing anyway at startup to see what's supported. IMO it's better to just leave the probing logic there and report one time. e.g. from my code:

```golang
	cell.Provide(func(l *slog.Logger) (sockets.SocketDestroyer, error) {
		bpfSD := &sockets.BPFSocketDestroyer{
			Logger: l,
		}

		if supported, err := bpfSD.IsSupported(); err != nil {
			return nil, fmt.Errorf("probing BPF socket destroyer support: %w", err)
		} else if supported {
			l.Info("Using BPF socket destroyer")
			return bpfSD, nil
		}

		l.Info("bpf_sock_destroy is not supported on the current kernel. Falling back to netlink-based socket destroyer")
		return &sockets.NetlinkSocketDestroyer{
			Logger: l,
		}, nil
	}),
```

Right now it just assumes the `NetlinkSocketDestroyer` will work, but it would be possible to add an `IsSupported()` function for that as well here. Worst case, this returns `nil` or an implementation of `SocketDestroyer` that does nothing and socket termination does nothing.

joamaki replied:
Yeah that makes sense to me. Would it be OK if you'd do that change in your PR once this is in?

jrife replied:
Yeah, sure. I will likely drop the `struct socketDestroyer` as when I rebase on this, since SocketDestroyer is `Provide`ed elsewhere. We can discuss more after merging this.

jrife reviewed (commented):
I cherry-picked my changes from https://github.com/cilium/cilium/pull/38693 on top of this branch and tested socket termination manually. Everything seemed to work; I saw the UDP socket get destroyed repeatedly.

aditighag commented on pkg/loadbalancer/reconciler/termination.go:
```diff
@@ -0,0 +1,252 @@
+
+			// Terminate the sockets connected to backends that have been either
+			// deleted or which are no longer considered viable.
+			if change.Deleted || !backend.IsAlive() {
```
Is it possible to get an update for a backend even when there was neither a change in the state nor it was deleted?

joamaki replied:
No it is not. The object in the change must have been either `Insert`ed or `Delete`d. Of course the contents of the object could be the same (even though it must be pointing to a different copy in the heap since it's a pointer (`Insert()` will panic otherwise)).

aditighag commented on pkg/loadbalancer/reconciler/termination_test.go:
```diff
@@ -0,0 +1,147 @@
+	wtxn.Commit()
+
+	// We should see two deletions: one for host ns and one for the mocked
+	// "foo" one.
```
Can we add two separate test cases -- (1) when socket-LB is enabled only in the host netns and verify that stale connections are terminated only in the host netns? (2) stale connections are terminated in both host netns and pod netns.

tommyp1ckles replied:
The [legacy](https://github.com/cilium/cilium/blob/main/pkg/loadbalancer/legacy/service/service_test.go#L2759-L2904) test also has coverage for checking that socket-cookie validation works.

joamaki replied:
I'll add that in as well.

joamaki replied:
Ported the test from service_test.go and I expanded the other test to cover the "only hostns" case as well.

aditighag commented on pkg/loadbalancer/maps/lbmaps.go:
```diff
@@ -762,6 +804,10 @@ func (f *FaultyLBMaps) DumpMaglev(cb func(lbmap.MaglevOuterKey, lbmap.MaglevOute
 	return f.impl.DumpMaglev(cb)
 }
 
+func (f *FaultyLBMaps) ExistsSockRevNat(cookie uint64, addr net.IP, port uint16) bool {
```
Umm... What's a faulty vs fake LB map? How is it relevant to this PR? Please consider splitting this change into a separate commit.

joamaki replied:
FaultyLBMaps is wrapping a `LBMaps` (either a fake or the real one) to provide fault injection for the tests. This way almost every test case we have exercises the retry machinery of the `pkg/loadbalancer/reconciler` as any BPF map operation can fail with some configurable probability. 

I cannot move it to a separate commit. Any addition to `LBMaps` must be implemented by `FakeLBMaps`, `BPFLBMaps` and by `FaultyLBMaps`. The code would fail to compile otherwise and if I move it to a separate commit the "build commits" CI job would fail.

aditighag reviewed (dismissed):
Sorry for the delay. Posted some comments from the first pass.
(1) Can you split the 2nd commit into preparatory changes (e.g., changes related to sock rev nat map creation/utility functions, etc) and changes porting the socket termination logic to the new LB control plane?
(2) How do we ensure the socket termination loop runs only on the relevant backend DB updates?

aditighag commented on pkg/loadbalancer/reconciler/termination.go:
```diff
@@ -0,0 +1,252 @@
+
+// netnsOps captures the operations performed with network namespaces.
+// This allows the tests to inject their own implementation.
+type netnsOps struct {
```
i like this approach for testing.

tommyp1ckles commented on pkg/loadbalancer/backend.go:
```diff
@@ -124,6 +124,21 @@ func (be *Backend) GetInstanceFromSource(name ServiceName, src source.Source) *B
 	return nil
 }
 
+// IsAlive returns true if any of the instances are marked active or terminanting and healthy.
```
typo: terminanting -> terminating

tommyp1ckles commented on pkg/loadbalancer/reconciler/termination_test.go:
```diff
@@ -0,0 +1,147 @@
+)
+
+func TestSocketTermination(t *testing.T) {
+	if testutils.IsPrivileged() {
```
Q: I don't see many tests doing this, should this just be something we tool as part of `testutils.Privileged()`?

joamaki replied:
What do you mean? "Tool" as in?

tommyp1ckles replied:
If we want to avoid running non-privileged tests on the privileged runner, to save on CI time then we should make that automatically happen, instead of having to manually set this up for each test.

joamaki replied:
Oh you mean I should've just used `testutils.PrivilegedTest()`? Yeah that makes sense. Changed.

tommyp1ckles commented on pkg/loadbalancer/reconciler/termination.go:
```diff
@@ -0,0 +1,252 @@
+		return err
+	}
+
+	// Process the changes in batches every 50 milliseconds.
```
Just as a point of discussion, if we're doing a batched termination approach - we could refactor the `Destroy` function to accept the whole batch of `SocketFilter`'s and then check for them all at once, avoiding unnecessary netlink iterations/message passing.

tommyp1ckles replied:
Probably makes sense as a follow up though.

joamaki replied:
Yeah we should, but not in this PR

joamaki committed:
loadbalancer/maps: Add SockRevNatMap

For implementing socket termination we'll need access to the SockRevNat BPF map.
Add it to the LBMaps interface and expand implementations.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
backend: Add IsAlive() method

Add IsAlive() for checking whether the backend is "alive", that is it has instances
that are either active & healthy, or if no active instances then a terminating one
that is healthy.

This will be used by follow-up commit on the socket termination to check if a backend
should be considered dead and thus the sockets connected to it should be terminated.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
loadbalancer/reconciler: Add benchmark for socket termination

Add benchmark for validating the overhead of processing every backend
change in the socket termination loop.

Results on my machine:

goos: linux
goarch: amd64
pkg: github.com/cilium/cilium/pkg/loadbalancer/reconciler
cpu: 13th Gen Intel(R) Core(TM) i9-13950HX
BenchmarkChangeIteration_TCP
BenchmarkChangeIteration_TCP-32               50          24211174 ns/op          41303242 backends/sec         24.21 ns/backend
BenchmarkChangeIteration_UDP
BenchmarkChangeIteration_UDP-32                5         247585791 ns/op           4039004 backends/sec        247.6 ns/backend
PASS

We're spending 24ns looking at each backend change. If we assume a busy cluster with say 100 backends
changing every 50 milliseconds we would be using 24ns * 100 * (1000/50) = 48000ns = 48us of CPU time every
second or in other words 100*(48000 / 10^9) = 0.0048% of a single CPU core.

If all changes were for UDP backends we'd hit the IsAlive() check and we'd get:
247ns * 100 * (1000/50) = 494us per second, which is 0.049% of a CPU core.

Increasing the rate limit interval to say 500ms or 1s would reduce the overhead of the revision and
graveyard revision index tree traversals significantly further reducing the CPU cost, but would increase
the latency at which we react to dead backends.

This seems like an acceptable trade-off to me considering that now the termination logic is completely
decoupled from everything else. The alternative way to implement this would be to have the BPF reconciler
queue up the terminations, but this would increase overall complexity.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
loadbalancer/reconciler: Datapath test of socket termination

Port the 'TestTerminateUDPConnectionsToBackend' from pkg/loadbalancer/legacy/service/service_test.go
to test the datapath side of the socket termination.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

dylandreimerink reviewed (approved):
Look good to me, any concerns were already brought up by other reviewers
